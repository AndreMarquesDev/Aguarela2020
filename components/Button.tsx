/* eslint-disable no-nested-ternary, jsx-a11y/anchor-is-valid */
import React, { FC } from 'react';
import Link from 'next/link';
import { useRouter } from 'next/router';
import classNames from 'classnames';
import { Page } from '../utils/pages';

interface ButtonProps {
    children: string;
    page?: Page;
    externalLink?: string;
    alignLeft?: boolean;
    isUppercased?: boolean;
    isSubmit?: boolean;
    onClick?: (event) => void;
}

const Button: FC<ButtonProps> = ({ children, page, externalLink, alignLeft, isUppercased, isSubmit, onClick }) => {
    const {
        query,
    } = useRouter();

    const language = query.language?.toString();
    const buttonBackgroundContainerStyles = classNames('buttonBackground', !alignLeft && 'alignCenter');
    const buttonStyles = classNames('button', isUppercased && 'uppercase');

    return (
        <>
            <div className="buttonWrapper">
                <div className={buttonBackgroundContainerStyles}>
                    {isSubmit ? (
                        <button className={buttonStyles} type="submit" onClick={onClick}>
                            {children}
                        </button>
                    ) : externalLink ? (
                        <a className={buttonStyles} href={externalLink} rel="noreferrer" target="_blank">
                            {children}
                        </a>
                    ) : (
                        <Link href={`/${language}/${page}`} prefetch={false}>
                            <a className={buttonStyles}>
                                {children}
                            </a>
                        </Link>

                    )}
                </div>
            </div>

            <style jsx>
                {`
                    @import './styles/_vars.scss';

                    .buttonWrapper {
                        width: 100%;
                        display: block;
                    }

                    .buttonBackground {
                        min-width: 180rem;
                        display: inline-block;
                        position: relative;
                        text-align: center;
                        margin-top: 50rem;
                        z-index: 0;
                        cursor: pointer;

                        &:hover {
                            &:before {
                                background-color: $pink;
                            }

                            &:after {
                                background-color: $blue;
                            }
                        }

                        &:before, &:after {
                            content: '';
                            width: 100%;
                            height: 100%;
                            position: absolute;
                            transition: background-color, .2s;
                        }

                        &:before {
                            top: 0;
                            left: 0;
                            background-color: $blue;
                        }

                        &:after {
                            top: 8rem;
                            left: 8rem;
                            background-color: $pink;
                            z-index: -1;
                        }

                        .button {
                            width: 100%;
                            position: relative;
                            display: block;
                            padding: 10rem;
                            @include fontS($yellow, uppercase, bold);
                            text-align: center;
                            outline: none;
                            z-index: 1;
                        }

                        .uppercase {
                            @include fontS($yellow, none, bold);

                            @include mobile {
                                @include fontXS($yellow, none, bold);
                            }
                        }
                    }

                    .alignCenter {
                        left: 50%;
                        transform: translateX(-50%);
                    }
                `}
            </style>
        </>
    );
};

export default Button;
