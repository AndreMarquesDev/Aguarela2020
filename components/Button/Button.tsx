/* eslint-disable no-nested-ternary, jsx-a11y/anchor-is-valid */
import React from 'react';
import Link from 'next/link';
import { useRouter } from 'next/router';
import classNames from 'classnames';
import { Page } from '../../utils/pages';

export interface ButtonProps {
    children: string;
    page?: Page;
    externalLink?: string;
    alignLeft?: boolean;
    isUppercased?: boolean;
    isSubmit?: boolean;
    disabled?: boolean;
    onClick?: (event) => void;
}

export const Button = ({
    children,
    page,
    externalLink,
    alignLeft,
    isUppercased,
    isSubmit,
    disabled,
    onClick,
}: ButtonProps): JSX.Element => {
    const { query } = useRouter();

    const language = query.language?.toString();
    const buttonBackgroundContainerStyles = classNames(
        'buttonBackground',
        !alignLeft && 'alignCenter',
        disabled && 'disabled'
    );
    const buttonStyles = classNames('button', isUppercased && 'uppercase');

    return (
        <>
            <div className="buttonWrapper">
                <div className={buttonBackgroundContainerStyles}>
                    {isSubmit ? (
                        <button className={buttonStyles} type="submit" onClick={onClick}>
                            {children}
                        </button>
                    ) : externalLink ? (
                        <a
                            className={buttonStyles}
                            href={externalLink}
                            rel="noreferrer"
                            target="_blank"
                        >
                            {children}
                        </a>
                    ) : (
                        <Link href={`/${language}/${page}`} prefetch={false}>
                            <a className={buttonStyles}>{children}</a>
                        </Link>
                    )}
                </div>
            </div>

            <style jsx>
                {`
                    @import './styles/_vars.scss';

                    .buttonWrapper {
                        width: 100%;
                        display: block;
                    }

                    .buttonBackground {
                        min-width: 180rem;
                        display: inline-block;
                        position: relative;
                        text-align: center;
                        margin-top: 50rem;
                        z-index: 0;
                        cursor: pointer;

                        &:hover {
                            &:before {
                                background-color: $pink;
                            }

                            &:after {
                                background-color: $blue;
                            }
                        }

                        &:before,
                        &:after {
                            content: '';
                            width: 100%;
                            height: 100%;
                            position: absolute;
                            transition: background-color, 0.2s;
                        }

                        &:before {
                            top: 0;
                            left: 0;
                            background-color: $blue;
                        }

                        &:after {
                            top: 8rem;
                            left: 8rem;
                            background-color: $pink;
                            z-index: -1;
                        }

                        .button {
                            width: 100%;
                            position: relative;
                            display: block;
                            padding: 10rem;
                            @include fontS($yellow, uppercase, bold);
                            text-align: center;
                            outline: none;
                            z-index: 1;
                        }

                        .uppercase {
                            @include fontS($yellow, none, bold);

                            @include mobile {
                                @include fontXS($yellow, none, bold);
                            }
                        }
                    }

                    .alignCenter {
                        left: 50%;
                        transform: translateX(-50%);
                    }

                    .disabled {
                        opacity: 0.5;
                        pointer-events: none;
                    }
                `}
            </style>
        </>
    );
};
