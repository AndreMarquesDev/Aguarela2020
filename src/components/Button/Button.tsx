import React from 'react';
import Link from 'next/link';
import { useRouter } from 'next/router';
import classNames from 'classnames';
import { Page } from '../../utils/pages';
import { mobileMediaQuery } from '../../styles/mediaQueries';

export interface ButtonProps {
    children: string;
    page?: Page;
    externalLink?: string;
    alignLeft?: boolean;
    isLowercased?: boolean;
    isSubmit?: boolean;
    disabled?: boolean;
    onClick?: () => void;
}

export const Button = ({
    children,
    page,
    externalLink,
    alignLeft,
    isLowercased,
    isSubmit,
    disabled,
    onClick,
}: ButtonProps): JSX.Element => {
    const { query } = useRouter();

    const language = query.language?.toString();
    const buttonBackgroundContainerStyles = classNames(
        'buttonBackground',
        !alignLeft && 'alignCenter',
        disabled && 'disabled'
    );
    const buttonStyles = classNames(
        'buttonComponent',
        'fontS',
        isLowercased && 'buttonComponentLowercase'
    );

    return (
        <>
            <div className="buttonWrapper">
                <div className={buttonBackgroundContainerStyles}>
                    {/* eslint-disable-next-line no-nested-ternary */}
                    {isSubmit ? (
                        <button className={buttonStyles} type="submit" onClick={onClick}>
                            {children}
                        </button>
                    ) : externalLink ? (
                        <a
                            className={buttonStyles}
                            href={externalLink}
                            rel="noreferrer"
                            target="_blank"
                        >
                            {children}
                        </a>
                    ) : (
                        <Link
                            className={buttonStyles}
                            href={`/${language}/${page}`}
                            prefetch={false}
                        >
                            {children}
                        </Link>
                    )}
                </div>
            </div>

            <style jsx>
                {`
                    .buttonWrapper {
                        width: 100%;
                        display: block;
                    }

                    .buttonBackground {
                        min-width: 180rem;
                        display: inline-block;
                        position: relative;
                        text-align: center;
                        margin-top: 50rem;
                        z-index: 0;
                        cursor: pointer;

                        &:hover {
                            &:before {
                                background-color: var(--pink);
                            }

                            &:after {
                                background-color: var(--blue);
                            }
                        }

                        &:before,
                        &:after {
                            content: '';
                            width: 100%;
                            height: 100%;
                            position: absolute;
                            transition: background-color, 0.2s;
                        }

                        &:before {
                            top: 0;
                            left: 0;
                            background-color: var(--blue);
                        }

                        &:after {
                            top: 8rem;
                            left: 8rem;
                            background-color: var(--pink);
                            z-index: -1;
                        }
                    }

                    .alignCenter {
                        left: 50%;
                        transform: translateX(-50%);
                    }

                    .disabled {
                        opacity: 0.5;
                        pointer-events: none;
                    }

                    :global(.buttonComponent) {
                        width: 100%;
                        position: relative;
                        display: block;
                        padding: 10rem;
                        color: var(--yellow);
                        font-weight: bold;
                        text-transform: uppercase;
                        text-align: center;
                        outline: none;
                        z-index: 1;
                    }

                    :global(.buttonComponentLowercase) {
                        text-transform: none;

                        @media (${mobileMediaQuery}) {
                            font-size: var(--fontSizeXS);
                        }
                    }
                `}
            </style>
        </>
    );
};
