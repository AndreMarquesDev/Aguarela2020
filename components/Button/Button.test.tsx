import '@testing-library/jest-dom';
import React from 'react';
import { render, screen } from '@testing-library/react';
import { Button, ButtonProps } from './Button';

const children = 'Texto dentro do bot√£o';

const baseProps: ButtonProps = {
    children,
    page: 'projects',
    externalLink: 'https://www.google.com',
    alignLeft: false,
    isUppercased: false,
    isSubmit: false,
    disabled: false,
    onClick: jest.fn(),
};

describe('<Button />', () => {
    test('renders properly', () => {
        const { container } = render(<Button {...baseProps}>{children}</Button>);

        const anchorElement = screen.getByText(children);

        expect(anchorElement).toHaveAttribute('href', baseProps.externalLink);
        expect(anchorElement).toHaveAttribute('target', '_blank');
        expect(screen.getByText(children)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test('renders properly without an external link', () => {
        const { container } = render(
            <Button {...baseProps} externalLink={null}>
                {children}
            </Button>
        );

        const anchorElement = screen.getByText(children);

        expect(anchorElement).toHaveAttribute('href', `/pt/${baseProps.page}`);
        expect(anchorElement).not.toHaveAttribute('target');
        expect(container).toMatchSnapshot();
    });

    test('renders properly as a submit button', () => {
        const { container } = render(
            <Button {...baseProps} isSubmit>
                {children}
            </Button>
        );

        const buttonElement = screen.getByText(children);

        expect(buttonElement).toHaveAttribute('type', 'submit');
        expect(container).toMatchSnapshot();
    });

    test('renders properly when disabled, uppercased and aligned left', () => {
        const { container } = render(
            <Button {...baseProps} alignLeft disabled isUppercased>
                {children}
            </Button>
        );

        expect(container).toMatchSnapshot();
    });
});
