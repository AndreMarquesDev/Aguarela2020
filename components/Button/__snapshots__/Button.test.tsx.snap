// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders properly 1`] = `
<div
  class="jsx-781296904 buttonWrapper"
>
  <div
    class="jsx-781296904 buttonBackground alignCenter"
    data-testid="button_container"
  >
    <a
      class="jsx-781296904 button"
      href="https://www.google.com"
      rel="noreferrer"
      target="_blank"
    >
      Texto dentro do bot達o
    </a>
  </div>
</div>
`;

exports[`<Button /> renders properly as a submit button 1`] = `
<div
  class="jsx-781296904 buttonWrapper"
>
  <div
    class="jsx-781296904 buttonBackground alignCenter"
    data-testid="button_container"
  >
    <button
      class="jsx-781296904 button"
      type="submit"
    >
      Texto dentro do bot達o
    </button>
  </div>
</div>
`;

exports[`<Button /> renders properly when disabled, uppercased and aligned left 1`] = `
<div
  class="jsx-781296904 buttonWrapper"
>
  <div
    class="jsx-781296904 buttonBackground disabled"
    data-testid="button_container"
  >
    <a
      class="jsx-781296904 button uppercase"
      href="https://www.google.com"
      rel="noreferrer"
      target="_blank"
    >
      Texto dentro do bot達o
    </a>
  </div>
</div>
`;

exports[`<Button /> renders properly without an external link 1`] = `
<div
  class="jsx-781296904 buttonWrapper"
>
  <div
    class="jsx-781296904 buttonBackground alignCenter"
    data-testid="button_container"
  >
    <a
      class="jsx-781296904 button"
      href="/pt/projects"
    >
      Texto dentro do bot達o
    </a>
  </div>
</div>
`;
